### PR TITLE
Add PDU session setup test

### LINKED ISSUE #28: Add tests for pdu session

### PR DESCRIPTION
**Pull request by Raguideau**
## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
Fix #28 
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

- [ ] I have updated the documentation accordingly.
- [x] I have added tests to cover my changes.
- [x] All new and existing tests passed.

- [x] I have read the **CONTRIBUTING** document.
- [x] Each of my commits messages include `Signed-off-by: Author Name <authoremail@example.com>` to accept the DCO.

### CODE REVIEW

File: test/aio5gc/build.go [hunk 1]
```diff
@@ -0,0 +1,59 @@
...
+	"my5G-RANTester/lib/ngap/ngapType"
+	"my5G-RANTester/test/aio5gc/context"
+	"my5G-RANTester/test/aio5gc/service"
+
+	"github.com/free5gc/nas"
+)
+
+type FiveGCBuilder struct {
+	config   config.Config
+	nasHook  func(*nas.Message, *context.UEContext, *context.Aio5gc) (bool, error)
```
Comment (linouxis9) [inline test/aio5gc/build.go:19]: could be a list of hook instead?

File: test/aio5gc/context/amf.go [hunk 2]
```diff
@@ -69,46 +80,42 @@ func (c *AMFContext) GetId() string {
...
 	for ue := range c.ues {
 		if c.ues[ue].amfNgapId == id {
-			ueMutex.Unlock()
 			return &c.ues[ue], nil
 		}
 	}
-	ueMutex.Unlock()
-
-	return &UEContext{}, errors.New("[AMF] UE with amfNgapId " + strconv.Itoa(int(id)) + "not found")
+	return &UEContext{}, errors.New("[5GC] UE with amfNgapId " + strconv.Itoa(int(id)) + "not found")
```
Comment (linouxis9) [inline test/aio5gc/context/amf.go:100]: return nil instead of empty &UEContext{}?

File: test/aio5gc/context/session.go [hunk 3]
```diff
@@ -0,0 +1,85 @@
...
+			Arp: &models.Arp{
+				PriorityLevel: 8,
+			},
+			PriorityLevel: 8,
+		},
+		SessRuleId: "SessRuleId-1",
+	}}
+
+	s.lastAllocatedIP = net.ParseIP("10.0.0.1")
+	s.n3 = net.ParseIP("127.0.0.1")
```
Comment (linouxis9) [inline test/aio5gc/context/session.go:59]: it's the UPF's ip address on N3?

File: test/aio5gc/context/sessionManagement.go [hunk 4]
```diff
@@ -0,0 +1,210 @@
...
+	case nasMessage.PDUSessionTypeIPv4:
+		addrLen = 4 + 1
+	case nasMessage.PDUSessionTypeIPv6:
+	case nasMessage.PDUSessionTypeIPv4IPv6:
+		addrLen = 12 + 1
+	}
+	return addr, addrLen
+}
+
+func CreatePDUSession(ulNasTransport *nasMessage.ULNASTransport,
```
Comment (linouxis9) [inline test/aio5gc/context/sessionManagement.go:123]: not sure if such function should be here

File: test/aio5gc/context/sessionManagement.go [hunk 5]
```diff
@@ -0,0 +1,210 @@
...
+	if m.GsmHeader.GetMessageType() != nas.MsgTypePDUSessionEstablishmentRequest {
+		return nil, errors.New("[5GC][NAS] UL NAS Transport container message expected to be PDU Session Establishment Request but was not")
+	}
+	sessionRequest := m.PDUSessionEstablishmentRequest
+
+	newSmContext.SetPti(sessionRequest.GetPTI())
+	newSmContext.SetPduSessionType(sessionRequest.GetPDUSessionTypeValue())
+	newSmContext.SetSessionRule(session.GetSessionRules()[0])
+	newSmContext.SetDefQosQFI(uint8(1))
+	ip, err := tools.IncrementIP(session.lastAllocatedIP.String(), "10.0.0.0/8")
```
Comment (linouxis9) [inline test/aio5gc/context/sessionManagement.go:179]: this is not thread-safe, add a dedicated method to update lastAllocatedIP and returns the allocated IP
you should try running the aio5gc against packetrusher multi-ue

File: test/aio5gc/lib/tools/tools.go [hunk 6]
```diff
@@ -46,9 +46,9 @@ func GenerateDefaultConf(controlIF config.ControlIF, dataIF config.DataIF, amf c
 				Nia2: true,
 			},
 			Ciphering: config.Ciphering{
-				Nea0: false,
+				Nea0: true,
```
Comment (linouxis9) [inline test/aio5gc/lib/tools/tools.go:49]: please keep NEA0 and NEA2 both at true
that way PacketRusher will work with 5G Core that only work with NEA0 by default (free5gc) and proprietary cores that mandates a secure ciphering algorithms by default

**Review by linouxis9 — APPROVED** [review]
Great! Could you share a PCAP?
Could you try with packetrusher multi-ue as well?
Thanks a lot for the PR!

**Comment by Raguideau** [comment]
Hi,
Here is the PCAP when running the test with 1 UE: [pduSetupTest.zip](https://github.com/HewlettPackard/PacketRusher/files/13587220/pduSetupTest.zip)
I updated GitHub Action workflow and the test so it runs with 10 UE now.

**Review by linouxis9 — APPROVED** [review]

**Comment by linouxis9** [comment]
> Hi, Here is the PCAP when running the test with 1 UE: [pduSetupTest.zip](https://github.com/HewlettPackard/PacketRusher/files/13587220/pduSetupTest.zip) I updated GitHub Action workflow and the test so it runs with 10 UE now.

Look all good to me, thanks a lot for your PR!

**Pull request merged**